Add FactoryVecDeque + FactoryHashMap builder APIs (and replace `forward_to_parent`)

chriskilding opened the pull request:
## Summary

This adds a builder for `FactoryVecDeque`, in the same style as the `SimpleComponent` builder.

The notable change is that the builder gives the user a choice about how to handle output events from the FactoryVecDeque's children:

- `forward_from_children(sender, transformer)`: transform output events using the specified transformer, then forward them to the specified sender
- `detach()`: ignore output events from children

In code, that looks like:

```rs
/// Adaptation of relm4/examples/factory.rs

fn init(counter, root, sender) -> ComponentParts<Self> {
    let counters = FactoryVecDeque::builder()
            .init(/* additional params here */)
            .forward_from_children(sender.input_sender(), |output| {
                match output {
                    CounterOutput::SendFront(index) => AppMsg::SendFront(index),
                    CounterOutput::MoveUp(index) => AppMsg::MoveUp(index),
                    CounterOutput::MoveDown(index) => AppMsg::MoveDown(index),
                    CounterOutput::Remove(index) => AppMsg::Remove(index),
                }
            });

}
```

This offers two main benefits:

- By using `forward_from_children(sender.input_sender(), ...)` instead of `forward_to_parent`, the child no longer requires knowledge of its parent to send events. This leads to a cleaner design.
- Most importantly, you can now do `.forward_from_children(sender.output_sender(), |msg| { ... })` to bubble a message from the child upwards to the next component. This is a critical feature to support document-based apps, which often have a complex document model (multiple nested elements), and a headless component at the top to handle reading/writing to the document state store.

## Checklist

- [x] cargo fmt
- [x] cargo clippy
- [x] cargo test
- [ ] updated CHANGES.md

## To do

- [x] FactoryVecDeque builder
- [x] FactoryHashMap builder
- [x] Update all examples
- [ ] Documentation

AaronErhardt commented:
Thanks for the PR! Let me know when this is ready for review or there's something I can do to support you.

This PR also got me wondering whether `init` might be a better method name for components as well instead of `launch`. In any case, this should probably be kept consistent.

chriskilding commented:
Great :)

The main thing I'm stuck on is the implementation of `detach` and `forward_from_children`. I don't know how to get the `sender` and `transform` arguments into the `FactoryVecDeque`, so that they can be added to its children. (Perhaps API changes to that class are also needed for this?)

Any and all help gratefully accepted, or if you want to jump in and edit the PR code that also works!

AaronErhardt commented:
Yes, it probably needs more changes. In fact, the `forward_to_parent` stuff must be removed entirely because we need the output receiver for forwarding. In general, this should be very similar then to how components are built. 

Maybe I have some time for working on the code today afternoon...

chriskilding commented:
I've reworked the code to include the changes from #526

Now I need to go update the examples and the documentation, but before that, a couple of thoughts...

chriskilding commented:
1. The design of `forward(sender, transform)` and `detach()` is intentionally similar to the SimpleComponent builder. I think it could be worth writing a common interface containing signatures for those methods, which we can then implement for the Connector, FactoryVecDequeBuilder, and FactoryHashMapBuilder. Thoughts?

chriskilding commented:
2. The SimpleComponent has a 2-stage builder: calling launch() on the Builder yields the Connector, and then you call forward() / detach() / etc on the Connector to yield the Component.

Meanwhile, the FactoryVecDeque / FactoryHashMap have only a 1-stage builder. This means that some arguments have to be supplied effectively as constructor arguments to the builder. This *might* limit the flexibility of the builder API in future.

For consistency I think that FactoryVecDeque / FactoryHashMap / SimpleComponent should have the same number of build stages. Should we introduce the Connector concept to FactoryVecDeque and FactoryHashMap as well?

AaronErhardt commented:
Regarding your thoughts:

1. Yes, adding shared traits is a good idea, but not necessary for this PR. It might make more sense to first write specific code and work on more generalization later.
2. Yes, consistency is definitely good and this would also probably help with archiving the goals of 1.

chriskilding commented:
I've updated the examples (at least, enough of them to make the build pass again) to the new FactoryVecDeque API.

I have also added a new example, `state_management`, which shows how to do persistent state management in a Relm4 app by loading and saving files. (But the same principles apply if your data store were something else, like a database or Web API.)

chriskilding commented:
I also intend to write up the `state_management` example as a new page in the Relm book, to help people who are new to the framework understand how to write apps with persistent state. (This is a task for later on, after this PR is complete)

chriskilding commented:
Note: this PR needs to be squashed before it is merged, to clean up the commits. Let me know at the point when it's good to go and I'll take care of that.

(Alternatively, you can use the squash merge option in this PR. I find this very helpful in my own projects to keep the main branch tidy.)

AaronErhardt commented on relm4/src/factory/sync/collections/hashmap.rs:
```diff
@@ -48,12 +49,105 @@ where
+where
+    C: FactoryComponent,
+{
+    pub fn forward<F, Msg>(self, f: F, forward_sender: Sender<Msg>) -> FactoryHashMap<K, C>
```
Currently, all `forward` methods on connectors seem to use `&Sender<Msg>` as parameter and it would be nice to keep this consistent.

chriskilding replied:
Agreed - I've updated this with that change. (Also I reordered the sender and the transformer arguments to match the order used in the FactoryVecDeque and Component connectors.)

AaronErhardt reviewed (approved):
Thanks a lot! This is really turning out to be a great change and makes Relm4 a lot more consistent overall.

I just noticed two small things to improve before merging.

chriskilding commented on relm4/src/factory/sync/collections/hashmap.rs:
```diff
@@ -85,33 +179,15 @@ where
 where
     C: FactoryComponent,
 {
-    /// Creates a new [`FactoryHashMap`].
-    #[must_use]
-    pub fn width_hasher(
-        widget: C::ParentWidget,
-        parent_sender: &Sender<C::ParentInput>,
-        hash_builder: S,
-    ) -> Self {
-        Self {
-            widget,
-            parent_sender: parent_sender.clone(),
-            inner: HashMap::with_hasher(hash_builder),
-        }
-    }
```
Aha - I couldn't find out how the width_hasher was used, so I had assumed it might have been unused code. I'll reinstate it 👍

chriskilding replied:
Also was it meant to be called `with_hasher` instead of `width_hasher`?

chriskilding commented:
The build is getting caught up on a mystery clippy problem, which wasn't happening a day or two ago...

The following section of code yields the Clippy error "this method could have a #[must_use] attribute":

```rs
/// Inner type of the data binding.
pub struct $name {
    #[property(get, set)]
    /// The primary value.
    value: RefCell<$ty>,

    // ...
```

But when I added that attribute as directed, it then yields the error "#[must_use] has no effect when applied to a struct field":

```rs
/// Inner type of the data binding.
pub struct $name {
    #[must_use]
    #[property(get, set)]
    /// The primary value.
    value: RefCell<$ty>,

    // ...
```

I'm not sure what's going on, but I think the source of this is from outside this PR.

Could you take a look at this and see if it can be fixed?

AaronErhardt commented:
Sometimes new clippy lints are merged into Rust nightly without extensive testing, so they often have some false positives (that's also why nightly is there in the first place, to detect such problems quickly). I'll look into it after work in the afternoon.

AaronErhardt commented:
Btw. you can ignore the CI error, it's caused by a regression in `glib-macros` from yesterday: https://github.com/gtk-rs/gtk-rs-core/issues/1186

Btw. I already had a `with_hasher` method on my branch, I think. You should be able to copy most code.

chriskilding commented:
Ok, I've reinstated the `hasher` method

AaronErhardt reviewed (approved):
Looks good, should we merge it or do you want to add anything else?

chriskilding commented:
I think it's all good for now

chriskilding commented:
(One thing - Shall I do the rebase, or do you want to do the squash merge?)

AaronErhardt commented:
Squash merge is faster 😉 

Thanks for your work on this! This will be one the highlights in 0.7 for sure 🚀

chriskilding commented:
Once the Clippy errors have cleared in `main`, would you be able to make a 0.7-alpha-2 release? This will enable me to start writing the book chapter on state management (with example code).
